File: src/main/java/io/github/oleksivio/telegram/bot/core/controller/annotations/AnnotationProcessor.java
package io.github.oleksivio.telegram.bot.core.controller.annotations;

import io.github.oleksivio.telegram.bot.api.annotations.behavior.Session;
import io.github.oleksivio.telegram.bot.core.controller.annotations.composer.SessionInitializer;
import io.github.oleksivio.telegram.bot.core.controller.annotations.composer.StatelessInitializer;
import org.springframework.stereotype.Controller;

@Controller
public class AnnotationProcessor {

    private final StatelessInitializer statelessInitializer;

    private final SessionInitializer sessionInitializer;

    public AnnotationProcessor(StatelessInitializer statelessInitializer,
                               SessionInitializer sessionInitializer) {
        this.statelessInitializer = statelessInitializer;
        this.sessionInitializer = sessionInitializer;
    }

    public void add(Object obj) {
        Class<?> objClz = obj.getClass();

        if (objClz.isAnnotationPresent(Session.class)) {
            sessionInitializer.init(objClz, obj);
        }
        else {
            statelessInitializer.init(objClz, obj);
        }
    }

}

File: src/main/kotlin/io/github/oleksivio/telegram/bot/core/controller/annotations/AnnotationProcessor.kt
package io.github.oleksivio.telegram.bot.core.controller.annotations

import io.github.oleksivio.telegram.bot.api.annotations.behavior.Session
import io.github.oleksivio.telegram.bot.core.controller.annotations.composer.SessionInitializer
import io.github.oleksivio.telegram.bot.core.controller.annotations.composer.StatelessInitializer
import org.springframework.stereotype.Controller
import kotlin.reflect.full.findAnnotation

@Controller
class AnnotationProcessor(private val statelessInitializer: StatelessInitializer,
                          private val sessionInitializer: SessionInitializer) {

    fun add(obj: Any) {
        val objClz = obj::class

        if (objClz.findAnnotation<Session>() != null) {
            sessionInitializer.init(objClz, obj)
        } else {
            statelessInitializer.init(objClz, obj)
        }
    }

}
